File: C++/base64_utils/base64_utils_test.cpp
/**
 * @file base64_utils_test.cpp
 * @brief Unit tests for Base64 Utilities
 *
 * Comprehensive test suite for Base64 encoding/decoding operations.
 *
 * @author AllToolkit Contributors
 * @version 1.0.0
 * @license MIT
 */

#include "mod.hpp"
#include <iostream>
#include <cassert>
#include <cstring>

using namespace alltoolkit;

// Test counter
int tests_run = 0;
int tests_passed = 0;

#define TEST(name) void test_##name()
#define RUN_TEST(name) do { \
    tests_run++; \
    try { \
        test_##name(); \
        std::cout << "  [PASS] " << #name << std::endl; \
        tests_passed++; \
    } catch (const std::exception& e) { \
        std::cout << "  [FAIL] " << #name << ": " << e.what() << std::endl; \
    } \
} while(0)

#define ASSERT_EQ(a, b) assert((a) == (b))
#define ASSERT_TRUE(x) assert((x))
#define ASSERT_FALSE(x) assert(!(x))

// =============================================================================
// Encoding Tests
// =============================================================================

TEST(encode_empty) {
    std::string result = Base64Utils::encode("");
    ASSERT_EQ(result, "");
}

TEST(encode_basic) {
    // RFC 4648 test vectors
    ASSERT_EQ(Base64Utils::encode("f"), "Zg==");
    ASSERT_EQ(Base64Utils::encode("fo"), "Zm8=");
    ASSERT_EQ(Base64Utils::encode("foo"), "Zm9v");
    ASSERT_EQ(Base64Utils::encode("foob"), "Zm9vYg==");
    ASSERT_EQ(Base64Utils::encode("fooba"), "Zm9vYmE=");
    ASSERT_EQ(Base64Utils::encode("foobar"), "Zm9vYmFy");
}

TEST(encode_hello_world) {
    ASSERT_EQ(Base64Utils::encode("Hello, World!"), "SGVsbG8sIFdvcmxkIQ==");
}

TEST(encode_binary_data) {
    std::vector<uint8_t> data = {0x00, 0x01, 0x02, 0x03, 0xFF};
    std::string result = Base64Utils::encode(data);
    ASSERT_EQ(result, "AAECA/8=");
}

TEST(encode_url_safe) {
    std::string input = "Hello+World/Test";
    std::string standard = Base64Utils::encode(input, false, true);
    std::string urlSafe = Base64Utils::encode(input, true, true);
    ASSERT_EQ(standard, "SGVsbG8rV29ybGQvVGVzdA==");
    ASSERT_EQ(urlSafe, "SGVsbG8rV29ybGQvVGVzdA==");
}

TEST(encode_no_padding) {
    ASSERT_EQ(Base64Utils::encode("f", false, false), "Zg");
    ASSERT_EQ(Base64Utils::encode("fo", false, false), "Zm8");
    ASSERT_EQ(Base64Utils::encode("foo", false, false), "Zm9v");
}

// =============================================================================
// Decoding Tests
// =============================================================================

TEST(decode_empty) {
    std::string result = Base64Utils::decode("");
    ASSERT_EQ(result, "");
}

TEST(decode_basic) {
    ASSERT_EQ(Base64Utils::decode("Zg=="), "f");
    ASSERT_EQ(Base64Utils::decode("Zm8="), "fo");
    ASSERT_EQ(Base64Utils::decode("Zm9v"), "foo");
    ASSERT_EQ(Base64Utils::decode("Zm9vYg=="), "foob");
    ASSERT_EQ(Base64Utils::decode("Zm9vYmE="), "fooba");
    ASSERT_EQ(Base64Utils::decode("Zm9vYmFy"), "foobar");
}

TEST(decode_hello_world) {
    ASSERT_EQ(Base64Utils::decode("SGVsbG8sIFdvcmxkIQ=="), "Hello, World!");
}

TEST(decode_binary_data) {
    std::vector<uint8_t> result = Base64Utils::decodeToBytes("AAECA/8=");
    ASSERT_EQ(result.size(), 5);
    ASSERT_EQ(result[0], 0x00);
    ASSERT_EQ(result[1], 0x01);
    ASSERT_EQ(result[2], 0x02);
    ASSERT_EQ(result[3], 0x03);
    ASSERT_EQ(result[4], 0xFF);
}

TEST(decode_no_padding) {
    ASSERT_EQ(Base64Utils::decode("Zg"), "f");
    ASSERT_EQ(Base64Utils::decode("Zm8"), "fo");
    ASSERT_EQ(Base64Utils::decode("Zm9v"), "foo");
}

TEST(decode_url_safe) {
    // URL-safe Base64 with - and _
    std::string result = Base64Utils::decode("SGVsbG8rV29ybGQvVGVzdA", true);
    ASSERT_EQ(result, "Hello+World/Test");
}

TEST(decode_invalid_char) {
    try {
        Base64Utils::decode("Zm9v!bar");
        assert(false);  // Should have thrown
    } catch (const std::invalid_argument&) {
        // Expected
    }
}

// =============================================================================
// Round-trip Tests
// =============================================================================

TEST(roundtrip_text) {
    std::string original = "The quick brown fox jumps over the lazy dog";
    std::string encoded = Base64Utils::encode(original);
    std::string decoded = Base64Utils::decode(encoded);
    ASSERT_EQ(decoded, original);
}

TEST(roundtrip_binary) {
    std::vector<uint8_t> original;
    for (int i = 0; i < 256; ++i) {
        original.push_back(static_cast<uint8_t>(i));
    }
    std::string encoded = Base64Utils::encode(original);
    std::vector<uint8_t> decoded = Base64Utils::decodeToBytes(encoded);
    ASSERT_EQ(original.size(), decoded.size());
    for (size_t i = 0; i < original.size(); ++i) {
        ASSERT_EQ(original[i], decoded[i]);
    }
}

TEST(roundtrip_url_safe) {
    std::string original = "Special chars: +/=";
    std::string encoded = Base64Utils::encode(original, true, false);
    std::string decoded = Base64Utils::decode(encoded, true);
    ASSERT_EQ(decoded, original);
}

// =============================================================================
// URL-safe Conversion Tests
// =============================================================================

TEST(to_url_safe) {
    ASSERT_EQ(Base64Utils::toUrlSafe("Hello+World/Test==", false), "Hello-World_Test");
    ASSERT_EQ(Base64Utils::toUrlSafe("Hello+World/Test==", true), "Hello-World_Test==");
}

TEST(from_url_safe) {
    ASSERT_EQ(Base64Utils::fromUrlSafe("Hello-World_Test"), "Hello+World/Test");
}

TEST(url_safe_roundtrip) {
    std::string standard = "SGVsbG8rV29ybGQvVGVzdA==";
    std::string urlSafe = Base64Utils::toUrlSafe(standard, true);
    std::string back = Base64Utils::fromUrlSafe(urlSafe);
    ASSERT_EQ(back, standard);
}

// =============================================================================
// Validation Tests
// =============================================================================

TEST(is_valid_empty) {
    ASSERT_TRUE(Base64Utils::isValid(""));
}

TEST(is_valid_standard) {
    ASSERT_TRUE(Base64Utils::isValid("SGVsbG8="));
    ASSERT_TRUE(Base64Utils::isValid("Zm9v"));
    ASSERT_TRUE(Base64Utils::isValid("Zg=="));
}

TEST(is_valid_url_safe) {
    ASSERT_TRUE(Base64Utils::isValid("SGVsbG8-", true));
    ASSERT_TRUE(Base64Utils::isValid("Zm9v_2E=", true));  // 8 chars with padding
}

TEST(is_valid_invalid) {
    ASSERT_FALSE(Base64Utils::isValid("Zm9v!"));
    ASSERT_FALSE(Base64Utils::isValid("Zm9v@"));
}

TEST(is_valid_wrong_length) {
    // Length % 4 == 1 is invalid
    ASSERT_FALSE(Base64Utils::isValid("Z"));
}

// =============================================================================
// Length Calculation Tests
// =============================================================================

TEST(encoded_length) {
    ASSERT_EQ(Base64Utils::encodedLength(0, true), 0);
    ASSERT_EQ(Base64Utils::encodedLength(1, true), 4);   // "Zg=="
    ASSERT_EQ(Base64Utils::encodedLength(2, true), 4);   // "Zm8="
    ASSERT_EQ(Base64Utils::encodedLength(3, true), 4);   // "Zm9v"
    ASSERT_EQ(Base64Utils::encodedLength(4, true), 8);   // "Zm9vYg=="
}

TEST(encoded_length_no_padding) {
    ASSERT_EQ(Base64Utils::encodedLength(1, false), 2);  // "Zg"
    ASSERT_EQ(Base64Utils::encodedLength(2, false), 3);  // "Zm8"
    ASSERT_EQ(Base64Utils::encodedLength(3, false), 4);  // "Zm9v"
}

TEST(decoded_max_length) {
    ASSERT_EQ(Base64Utils::decodedMaxLength(0), 0);
    ASSERT_EQ(Base64Utils::decodedMaxLength(4), 3);
    ASSERT_EQ(Base64Utils::decodedMaxLength(8), 6);
    ASSERT_EQ(Base64Utils::decodedMaxLength(5), 3);
}

// =============================================================================
// Unicode Tests
// =============================================================================

TEST(encode_unicode) {
    std::string chinese = "Hello World!";
    std::string encoded = Base64Utils::encode(chinese);
    std::string decoded = Base64Utils::decode(encoded);
    ASSERT_EQ(decoded, chinese);
}

TEST(encode_multibyte) {
    std::string emoji = "Hello World!";
    std::string encoded = Base64Utils::encode(emoji);
    std::string decoded = Base64Utils::decode(encoded);
    ASSERT_EQ(decoded, emoji);
}

// =============================================================================
// Main
// =============================================================================

int main() {
    std::cout << "==============================================" << std::endl;
    std::cout << "Base64 Utilities Test Suite" << std::endl;
    std::cout << "==============================================" << std::endl;

    std::cout << "\n--- Encoding Tests ---" << std::endl;
    RUN_TEST(encode_empty);
    RUN_TEST(encode_basic);
    RUN_TEST(encode_hello_world);
    RUN_TEST(encode_binary_data);
    RUN_TEST(encode_url_safe);
    RUN_TEST(encode_no_padding);

    std::cout << "\n--- Decoding Tests ---" << std::endl;
    RUN_TEST(decode_empty);
    RUN_TEST(decode_basic);
    RUN_TEST(decode_hello_world);
    RUN_TEST(decode_binary_data);
    RUN_TEST(decode_no_padding);
    RUN_TEST(decode_url_safe);
    RUN_TEST(decode_invalid_char);

    std::cout << "\n--- Round-trip Tests ---" << std::endl;
    RUN_TEST(roundtrip_text);
    RUN_TEST(roundtrip_binary);
    RUN_TEST(roundtrip_url_safe);

    std::cout << "\n--- URL-safe Conversion Tests ---" << std::endl;
    RUN_TEST(to_url_safe);
    RUN_TEST(from_url_safe);
    RUN_TEST(url_safe_roundtrip);

    std::cout << "\n--- Validation Tests ---" << std::endl;
    RUN_TEST(is_valid_empty);
    RUN_TEST(is_valid_standard);
    RUN_TEST(is_valid_url_safe);
    RUN_TEST(is_valid_invalid);
    RUN_TEST(is_valid_wrong_length);

    std::cout << "\n--- Length Calculation Tests ---" << std::endl;
    RUN_TEST(encoded_length);
    RUN_TEST(encoded_length_no_padding);
    RUN_TEST(decoded_max_length);

    std::cout << "\n--- Unicode Tests ---" << std::endl;
    RUN_TEST(encode_unicode);
    RUN_TEST(encode_multibyte);

    std::cout << "\n==============================================" << std::endl;
    std::cout << "Results: " << tests_passed << "/" << tests_run << " tests passed" << std::endl;
    std::cout << "==============================================" << std::endl;

    return (tests_passed == tests_run) ? 0 : 1;
}

File: Lua/base64_utils/base64_utils_test.lua
--[[
Base64 Utils 测试文件
测试所有 Base64 工具函数
]]

-- 加载模块
package.path = package.path .. ";../?.lua"
local base64 = require("mod")

-- 测试计数
local tests_passed = 0
local tests_failed = 0

-- 简单测试框架
local function test(name, func)
    local success, err = pcall(func)
    if success then
        tests_passed = tests_passed + 1
        print("✓ " .. name)
    else
        tests_failed = tests_failed + 1
        print("✗ " .. name .. ": " .. tostring(err))
    end
end

local function assert_eq(actual, expected, msg)
    if actual ~= expected then
        error((msg or "") .. " 期望: " .. tostring(expected) .. " 实际: " .. tostring(actual))
    end
end

local function assert_contains(str, substr, msg)
    if not string.find(str, substr, 1, true) then
        error((msg or "") .. " 字符串应包含: " .. substr)
    end
end

print("========== Base64 Utils 测试开始 ==========\n")

-- 测试基本编码
test("基本编码 - 空字符串", function()
    local result = base64.encode("")
    assert_eq(result, "", "空字符串编码")
end)

test("基本编码 - Hello", function()
    local result = base64.encode("Hello")
    assert_eq(result, "SGVsbG8=", "Hello 编码")
end)

test("基本编码 - Hello World", function()
    local result = base64.encode("Hello World")
    assert_eq(result, "SGVsbG8gV29ybGQ=", "Hello World 编码")
end)

test("基本编码 - 数字", function()
    local result = base64.encode("1234567890")
    assert_eq(result, "MTIzNDU2Nzg5MA==", "数字编码")
end)

test("基本编码 - 特殊字符", function()
    local result = base64.encode("!@#$%^&*()")
    assert_eq(result, "IUAjJCVeJiooKQ==", "特殊字符编码")
end)

test("基本编码 - 中文", function()
    local result = base64.encode("你好世界")
    assert_eq(result, "5L2g5aW95LiW55WM", "中文编码")
end)

-- 测试基本解码
test("基本解码 - 空字符串", function()
    local result, err = base64.decode("")
    assert_eq(err, nil, "无错误")
    assert_eq(result, "", "空字符串解码")
end)

test("基本解码 - Hello", function()
    local result, err = base64.decode("SGVsbG8=")
    assert_eq(err, nil, "无错误")
    assert_eq(result, "Hello", "Hello 解码")
end)

test("基本解码 - Hello World", function()
    local result, err = base64.decode("SGVsbG8gV29ybGQ=")
    assert_eq(err, nil, "无错误")
    assert_eq(result, "Hello World", "Hello World 解码")
end)

test("基本解码 - 中文", function()
    local result, err = base64.decode("5L2g5aW95LiW55WM")
    assert_eq(err, nil, "无错误")
    assert_eq(result, "你好世界", "中文解码")
end)

-- 测试编解码往返
test("编解码往返 - 随机数据", function()
    local original = "This is a test string with various characters: 123!@#"
    local encoded = base64.encode(original)
    local decoded, err = base64.decode(encoded)
    assert_eq(err, nil, "无错误")
    assert_eq(decoded, original, "往返一致")
end)

test("编解码往返 - 二进制数据", function()
    local original = string.char(0, 1, 2, 127, 128, 255, 254, 253)
    local encoded = base64.encode(original)
    local decoded, err = base64.decode(encoded)
    assert_eq(err, nil, "无错误")
    assert_eq(decoded, original, "二进制往返一致")
end)

-- 测试 URL 安全编码
test("URL安全编码 - 基本测试", function()
    local result = base64.encode_urlsafe("Hello World")
    assert_eq(result, "SGVsbG8gV29ybGQ", "无 padding")
    assert_eq(string.find(result, "="), nil, "无等号")
end)

test("URL安全编码 - 特殊字符转换", function()
    -- 包含会生成 +/ 的数据
    local data = string.char(0xfb, 0xff)
    local result = base64.encode_urlsafe(data)
    assert_eq(string.find(result, "+"), nil, "无加号")
    assert_eq(string.find(result, "/"), nil, "无斜杠")
end)

test("URL安全解码 - 基本测试", function()
    local original = "Test Data!"
    local encoded = base64.encode_urlsafe(original)
    local decoded, err = base64.decode_urlsafe(encoded)
    assert_eq(err, nil, "无错误")
    assert_eq(decoded, original, "URL安全往返一致")
end)

-- 测试验证
test("验证 - 有效 Base64", function()
    assert_eq(base64.is_valid("SGVsbG8="), true, "有效标准格式")
    assert_eq(base64.is_valid("SGVsbG8gV29ybGQ="), true, "有效长格式")
end)

test("验证 - 无效 Base64", function()
    assert_eq(base64.is_valid("SGVs=bG8="), false, "无效 padding 位置")
    assert_eq(base64.is_valid("SGVs##G8="), false, "含非法字符")
    assert_eq(base64.is_valid(""), false, "空字符串")
end)

test("验证 - 含空格的 Base64", function()
    -- 注意：is_valid 会忽略空白字符（与解码行为一致）
    assert_eq(base64.is_valid("SGVs bG8="), true, "含空格仍然有效")
end)

test("验证 - URL 安全格式", function()
    assert_eq(base64.is_valid("SGVsbG8-_"), true, "URL安全字符有效")
end)

-- 测试统计
test("统计信息 - 基本测试", function()
    local stats = base64.get_encode_stats("Hello")
    assert_eq(stats.original_size, 5, "原始大小")
    assert_eq(stats.encoded_size, 8, "编码大小")
    assert_eq(stats.padding_count, 1, "padding 数量")
end)

test("统计信息 - 无 padding", function()
    local stats = base64.get_encode_stats("Hell")
    assert_eq(stats.original_size, 4, "原始大小")
    assert_eq(stats.padding_count, 2, "padding 数量")
end)

test("统计信息 - 不需要 padding", function()
    local stats = base64.get_encode_stats("HelloHelloHello")
    assert_eq(stats.original_size, 15, "原始大小")
    assert_eq(stats.padding_count, 0, "无 padding")
end)

-- 测试表格编解码
test("表格编码 - 简单表格", function()
    local tbl = { name = "test", value = 123 }
    local encoded, err = base64.encode_table(tbl)
    assert_eq(err, nil, "无错误")
    assert_eq(type(encoded), "string", "返回字符串")
    assert_eq(#encoded > 0, true, "非空")
end)

test("表格编码 - 嵌套表格", function()
    local tbl = {
        user = {
            name = "Alice",
            age = 30,
            tags = { "dev", "lua" }
        }
    }
    local encoded, err = base64.encode_table(tbl)
    assert_eq(err, nil, "无错误")
end)

test("表格编码 - 数组", function()
    local tbl = { 1, 2, 3, "four", true }
    local encoded, err = base64.encode_table(tbl)
    assert_eq(err, nil, "无错误")
end)

test("表格解码 - 简单表格", function()
    local original = { name = "test", count = 42 }
    local encoded, _ = base64.encode_table(original)
    local decoded, err = base64.decode_table(encoded)
    assert_eq(err, nil, "无错误")
    assert_eq(decoded.name, "test", "name 字段")
    assert_eq(decoded.count, 42, "count 字段")
end)

test("表格解码 - 嵌套表格", function()
    local original = {
        config = {
            host = "localhost",
            port = 8080
        },
        items = { "a", "b", "c" }
    }
    local encoded, _ = base64.encode_table(original)
    local decoded, err = base64.decode_table(encoded)
    assert_eq(err, nil, "无错误")
    assert_eq(decoded.config.host, "localhost", "嵌套字段")
    assert_eq(#decoded.items, 3, "数组长度")
end)

-- 测试边界情况
test("边界 - 单字符", function()
    local original = "A"
    local encoded = base64.encode(original)
    local decoded, _ = base64.decode(encoded)
    assert_eq(decoded, original, "单字符往返")
end)

test("边界 - 两字符", function()
    local original = "AB"
    local encoded = base64.encode(original)
    local decoded, _ = base64.decode(encoded)
    assert_eq(decoded, original, "两字符往返")
end)

test("边界 - 大数据", function()
    -- 生成 10KB 数据
    local original = string.rep("TestData123", 1000)
    local encoded = base64.encode(original)
    local decoded, err = base64.decode(encoded)
    assert_eq(err, nil, "无错误")
    assert_eq(decoded, original, "大数据往返")
end)

test("边界 - 空白字符处理", function()
    local encoded = "SGVs\nbG8=\t "
    local decoded, err = base64.decode(encoded)
    assert_eq(err, nil, "无错误")
    assert_eq(decoded, "Hello", "空白字符被忽略")
end)

-- 测试错误处理
test("错误处理 - 无效输入类型", function()
    local success, err = pcall(function()
        base64.encode(123)
    end)
    assert_eq(success, false, "应该报错")
    assert_contains(err, "字符串", "错误信息包含类型提示")
end)

test("错误处理 - 解码无效字符", function()
    local decoded, err = base64.decode("SGVs###G8=")
    assert_eq(decoded, nil, "返回 nil")
    assert_eq(err ~= nil, true, "有错误信息")
end)

test("错误处理 - 解码无效长度", function()
    -- 7字符（非4的倍数，无padding）应该被处理（不崩溃）
    local decoded, err = base64.decode("SGVsbG8")
    -- 标准格式下，非4倍数长度可能报错或尝试解码
    -- 主要确保函数不崩溃
    assert_eq(decoded == nil or decoded ~= nil, true, "不崩溃")
end)

-- 测试标准测试向量
test("标准测试向量 - RFC 4648", function()
    -- RFC 4648 测试向量
    local vectors = {
        { "", "" },
        { "f", "Zg==" },
        { "fo", "Zm8=" },
        { "foo", "Zm9v" },
        { "foob", "Zm9vYg==" },
        { "fooba", "Zm9vYmE=" },
        { "foobar", "Zm9vYmFy" },
    }
    
    for _, v in ipairs(vectors) do
        local encoded = base64.encode(v[1])
        assert_eq(encoded, v[2], "编码: " .. v[1])
        
        local decoded, _ = base64.decode(v[2])
        assert_eq(decoded, v[1], "解码: " .. v[2])
    end
end)

-- 打印结果
print("\n========== 测试结果 ==========")
print(string.format("通过: %d", tests_passed))
print(string.format("失败: %d", tests_failed))
print(string.format("总计: %d", tests_passed + tests_failed))
print("========== 测试结束 ==========")

-- 返回状态
os.exit(tests_failed > 0 and 1 or 0)